### PR TITLE
ddl: integrate add index operator to dist framework

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #46258

Problem Summary:

Before this PR, add-index operators do not support row count statistic and metrics.

### What is changed and how it works?

Row count statistic:
- Remove the row count mechanism that depends on etcd.
- Save the row count in subtask meta.
- Collect subtask metas periodically and update `job.RowCount`.

Row count metric:
- Add a metric counter in `indexWriteResultSink`.

Refactor:
- Move flush logic into `indexWriteResultSink`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  ```sql
  mysql> admin show ddl jobs;
  +--------+---------+------------+------------------------+----------------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+
  | JOB_ID | DB_NAME | TABLE_NAME | JOB_TYPE               | SCHEMA_STATE         | SCHEMA_ID | TABLE_ID | ROW_COUNT | CREATE_TIME         | START_TIME          | END_TIME            | STATE         |
  +--------+---------+------------+------------------------+----------------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+
  |    330 | test    | sbtest1    | add index /* ingest */ | public               |       278 |      328 |  10000000 | 2023-08-30 17:26:22 | 2023-08-30 17:26:22 | 2023-08-30 17:26:33 | synced        |
  ...
  |    320 | test    | sbtest1    | add index /* ingest */ | public               |       278 |      307 |         0 | 2023-08-30 17:02:32 | 2023-08-30 17:02:32 | 2023-08-30 17:04:16 | synced        |
  +--------+---------+------------+------------------------+----------------------+-----------+----------+-----------+---------------------+---------------------+---------------------+---------------+
  10 rows in set (0.01 sec)
  
  mysql> select @@tidb_enable_dist_task;
  +-------------------------+
  | @@tidb_enable_dist_task |
  +-------------------------+
  |                       1 |
  +-------------------------+
  1 row in set (0.00 sec)
  ```
  Before this PR(JOB_ID = 320), the row count is not recorded. After this PR(JOB_ID = 330), we can see the row count.
  img_v2_e4563513-7602-47eb-aefe-38904439a0bg

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/backfilling_operator.go [hunk 1]
```diff
@@ -505,6 +534,24 @@ func (s *indexWriteResultSink) collectResult() error {
...
+	if s.tbl.GetPartitionedTable() != nil {
+		flushMode = ingest.FlushModeForceGlobal
+	}
+	idxInfo := s.index.Meta()
+	_, _, err := s.backendCtx.Flush(idxInfo.ID, flushMode)
+	if err != nil {
+		if common.ErrFoundDuplicateKeys.Equal(err) {
+			err = convertToKeyExistsErr(err, idxInfo, s.tbl.Meta())
+		}
+		logutil.BgLogger().Error("flush error",
```
Comment (wjhuang2016) [inline ddl/backfilling_operator.go:548]: The level shouldn't be error for duplicated key.

File: ddl/index.go [hunk 2]
```diff
@@ -1922,7 +1928,39 @@ func (w *worker) executeDistGlobalTask(reorgInfo *reorgInfo) error {
...
 	})
-	return g.Wait()
+	err = g.Wait()
+	return err
+}
+
+func (w *worker) updateJobRowCount(taskKey string, jobID int64) {
+	taskMgr, err := storage.GetTaskManager()
+	if err != nil {
+		logutil.BgLogger().Error("cannot get task manager", zap.String("category", "ddl"), zap.String("task_key", taskKey), zap.Error(err))
```
Comment (wjhuang2016) [inline ddl/index.go:1938]: Warn

File: ddl/index.go [hunk 3]
```diff
@@ -1922,7 +1928,39 @@ func (w *worker) executeDistGlobalTask(reorgInfo *reorgInfo) error {
...
+		return
+	}
+	gTask, err := taskMgr.GetGlobalTaskByKey(taskKey)
+	if err != nil || gTask == nil {
+		logutil.BgLogger().Error("cannot get global task", zap.String("category", "ddl"), zap.String("task_key", taskKey), zap.Error(err))
+		return
+	}
+	metas, err := taskMgr.GetSucceedSubtaskMeta(gTask.ID)
```
Comment (ywqzzy) [inline ddl/index.go:1946]: Can we add a field in `tidb_background_subtask` to record the rowCount the subtask processed?
Then the marshal and unmarshal and count rowCount process will be eliminated.
We can discussion it later.

File: disttask/framework/storage/task_table.go [hunk 4]
```diff
@@ -387,6 +387,24 @@ func (stm *TaskManager) GetSucceedSubtasksByStep(taskID int64, step int64) ([]*p
 	return subtasks, nil
 }
 
+// GetSucceedSubtaskMeta gets the subtask meta in the success state.
+func (stm *TaskManager) GetSucceedSubtaskMeta(taskID int64) ([][]byte, error) {
+	rs, err := stm.executeSQLWithNewSession(stm.ctx, `select meta from mysql.tidb_background_subtask
```
Comment (wjhuang2016) [inline disttask/framework/storage/task_table.go:392]: step is not limited, we may get the other step.

File: disttask/framework/storage/table_test.go [hunk 5]
```diff
@@ -224,11 +224,19 @@ func TestSubTaskTable(t *testing.T) {
 	subtasks, err := sm.GetSucceedSubtasksByStep(2, proto.StepInit)
 	require.NoError(t, err)
 	require.Len(t, subtasks, 0)
+	subtaskMetas, err := sm.GetSucceedSubtaskMeta(2)
```
Comment (okJiang) [inline disttask/framework/storage/table_test.go:227]: Can you use a variable `taskIDxxx` to represent 2

File: disttask/framework/storage/table_test.go [hunk 6]
```diff
@@ -224,11 +224,19 @@ func TestSubTaskTable(t *testing.T) {
...
+
 	err = sm.FinishSubtask(2, []byte{})
 	require.NoError(t, err)
+
 	subtasks, err = sm.GetSucceedSubtasksByStep(2, proto.StepInit)
 	require.NoError(t, err)
 	require.Len(t, subtasks, 1)
+	subtaskMetas, err = sm.GetSucceedSubtaskMeta(2)
+	require.NoError(t, err)
+	require.Len(t, subtaskMetas, 1)
```
Comment (okJiang) [inline disttask/framework/storage/table_test.go:239]: Can you add a test for `subtaskMetas[0]`?

File: tests/realtikvtest/addindextest/operator_test.go [hunk 7]
```diff
@@ -217,6 +224,7 @@ func TestBackfillOperatorPipeline(t *testing.T) {
 	require.NoError(t, err)
 	require.Len(t, keys, 10)
 	require.Len(t, values, 10)
+	require.Equal(t, int64(10), totalRowCount.Load())
```
Comment (okJiang) [inline tests/realtikvtest/addindextest/operator_test.go:227]: Does this line mean we can remove L225-L226?

File: ddl/index.go (hunk 3 above)
Comment (tangenta) [reply ddl/index.go:1946]: https://github.com/pingcap/tidb/pull/46562

File: disttask/framework/storage/table_test.go (hunk 5 above)
Comment (tangenta) [reply disttask/framework/storage/table_test.go]: Then we need to change the whole file and bring unnecessary LOC... I prefer to do this in another PR.

File: tests/realtikvtest/addindextest/operator_test.go (hunk 7 above)
Comment (tangenta) [reply tests/realtikvtest/addindextest/operator_test.go:220]: Removed.

File: ddl/backfilling_operator.go (hunk 1 above)
Comment (tangenta) [reply ddl/backfilling_operator.go:549]: Updated.

**Review by wjhuang2016 — APPROVED** [review]

File: ddl/ddl.go [hunk 8]
```diff
@@ -681,13 +681,13 @@ func newDDL(ctx context.Context, options ...Option) *ddl {
...
-		})
+		func(ctx context.Context, task *proto.Task, summary *scheduler.Summary) (scheduler.Scheduler, error) {
+			return NewBackfillSchedulerHandle(ctx, task.Meta, d, task.Step == proto.StepTwo, summary)
+		}, scheduler.WithSummary)
 
 	scheduler.RegisterSchedulerConstructor("backfill", proto.StepTwo,
-		func(ctx context.Context, _ int64, taskMeta []byte, step int64) (scheduler.Scheduler, error) {
-			return NewBackfillSchedulerHandle(ctx, taskMeta, d, step == proto.StepTwo)
+		func(ctx context.Context, task *proto.Task, summary *scheduler.Summary) (scheduler.Scheduler, error) {
+			return NewBackfillSchedulerHandle(ctx, task.Meta, d, task.Step == proto.StepTwo, nil)
```
Comment (Benjamin2037) [inline ddl/ddl.go:690]: why do we still has task.step and proto.StepTwo exist?

**Review by okJiang — COMMENTED** [review]
lgtm

**Review by okJiang — APPROVED** [review]

**Pull request merged**